### PR TITLE
ESLint: Fix `jest-dom/prefer-to-have-text-content` rule violations

### PR DESCRIPTION
**Pull request by tyxla**
## What?
This PR fixes all violations of the [`jest-dom/prefer-to-have-text-content` rule](https://github.com/testing-library/eslint-plugin-jest-dom/blob/main/docs/rules/prefer-to-have-text-content.md), in preparation for enabling `eslint-plugin-jest-dom` globally for the project.

See the [plugin README](https://github.com/testing-library/eslint-plugin-jest-dom) for more info on the jest-dom ESLint plugin.

## Why?
We've been improving our tests a lot recently with the migration to `@testing-library`. One way we could improve them is to better standardize them and follow best practices whenever we can, and one of the best ways to get there is to follow the recommended ESLint rules of the libraries and utilities we use under the hood.

## How?
We're essentially fixing up a few instances that either assert positively or negatively that the element has a certain text content.

## Testing Instructions
Verify all checks are green.

### CODE REVIEW

**Review by ciampo — APPROVED** [review]

**Comment by gziolo** [comment]
Where is the issue where you discuss introducing `eslint-plugin-jest-dom` plugin in the first place? 😄 

I briefly checked a few open PRs, and I don't think it really prevents "common mistakes when writing tests with jest-dom". All tests work exactly like before. It just adds syntax sugar on top of what you can do with Jest out of the box.

In fact, in this PR `toHaveTextContent` probably uses a less strict check as it works when refactoring:

```js
expect( el.textContent ).toEqual(
	expect.stringContaining( highlight )
);
```

to 

```js
expect( el ).toHaveTextContent( highlight );
```

Do you really feel that we should force all contributors to use only assertions from the plugin?

> One way we could improve them is to better standardize them and follow best practices whenever we can, and one of the best ways to get there is to follow the recommended ESLint rules of the libraries and utilities we use under the hood.

I'm a little bit worried that those best practices are too opinionated as they favor RTL custom assertions that are only one subjective way of writing code. If you really feel strongly about using standards, what would be the way to onboard new contributors so they don't have to fight with the linter to write tests?

**Comment by tyxla** [comment]
> Where is the issue where you discuss introducing eslint-plugin-jest-dom plugin in the first place? 😄

We'll have a PR soon - stay tuned.

> I briefly checked a few open PRs, and I don't think it really prevents "common mistakes when writing tests with jest-dom". All tests work exactly like before. It just adds syntax sugar on top of what you can do with Jest out of the box.

Agreed. It's mostly about test code consistency and using the most eloquent and readable ways to assert something about the DOM. I believe this improves code style as a whole, and that affects developer experience in a positive way.

> Do you really feel that we should force all contributors to use only assertions from the plugin?

This is only about a few specific DOM matchers, and I believe it's a more readable way to write those assertions, so why not follow that as a best practice, and have the linter help with it?

> I'm a little bit worried that those best practices are too opinionated as they favor RTL custom assertions that are only one subjective way of writing code.

I don't share your concerns here, to be honest. This is in line with one of WordPress's core philosophies - [Decisions, not options](https://wordpress.org/about/philosophy/#decisions). I believe having IDE to aid you with what assertion to write, and having clearer errors when а test is failing is a much better dev experience than what we had before with direct DOM traversing and manipulation. Anyway, I have a [branch](add/eslint-plugin-jest-dom) for enabling it and I'll make sure to ping you once I have a PR for it. It will be a good place to discuss the pros and cons, weigh the trade-offs, and raise your concerns.

> If you really feel strongly about using standards, what would be the way to onboard new contributors so they don't have to fight with the linter to write tests?

TBH, the best way to onboard new contributors into writing tests is to have a consistent test codebase, which we're achieving by following `jest-dom` best practices more closely. New contributors will often jump into existing tests to see how things are done, and if they're done in a consistent way, it will make things easier, rather than if we have 5 different ways to solve a problem. That's not to say that we can't work to improve developer docs - of course, we can, and I believe we should - providing more examples and explaining the concepts, and generally expanding the developer docs is something we should continuously be working on.

**Comment by gziolo** [comment]
>  I believe this improves code style as a whole, and that affects developer experience in a positive way.

I think it mostly depends on whether you are deeply familiar with RTL in the first place. Note that developers need to refer to the documentation pages on two different projects when using custom assertions. That's why I'm a little bit worried about less experienced contributors. Keep in mind that there are nearly 1 000 folks that contributed to the project, and they often aren't deeply familiar with the latest trends in the JavaScript world.

> This is in line with one of WordPress's core philosophies - [Decisions, not options](https://wordpress.org/about/philosophy/#decisions).

This one applies to the end users, but it's often used to make the point when talking about code, too 😄

**Comment by Mamaduka** [comment]
We have ["Best Practices" section](https://github.com/WordPress/gutenberg/tree/trunk/test/e2e#best-practices) for Playwright e2e tests.

If similar best practices are encouraged by linter, I think that would make reviewing PRs that include tests easier.

**Comment by tyxla** [comment]
> I think it mostly depends on whether you are deeply familiar with RTL in the first place. Note that developers need to refer to the documentation pages on two different projects when using custom assertions. That's why I'm a little bit worried about less experienced contributors. Keep in mind that there are nearly 1 000 folks that contributed to the project, and they often aren't deeply familiar with the latest trends in the JavaScript world.

I actually value that feedback - it's really useful! And I do understand your concern, but I believe it's way more flat learning curve than having to Learn `@testing-library`, or `jest`, or even `redux` and `react` themselves. And let me actually explain why I think it's even less complex now than it was before. 

When someone types `element.textContent` in their test, they get a very helpful message that they should be using `toHaveTextContent()` instead. I believe this (and the rest of the matchers) won't cause any friction, because they're so straightforward to replace, and even come quite naturally once you've written one or two. For example, it's way more natural to assert that something is `.toBeChecked()` rather than having to manually check for `checked` attribute and then the `aria-checked` attribute as well. What about checking if an element is focused? A new contributor could never think of using `document.activeElement`, but `.toHaveFocus()` is so straightforward. What about an input being disabled? Should I check if the attribute `disabled` exists? Should it have a value or not? What should the value be? It's so much easier to just go with `.toBeDisabled()` and get over with it. And I can continue with the examples, there's plenty more 😉 

To summarize, I actually believe that by removing Enzyme we made things simple for newer developers. IMO it would be much more confusing when you see 2 different flavors of component tests, rather than just one. It's the same with assertions and jest matchers and ways to query for elements and navigate between them - when you have less ways to do something, it quickly becomes more straightforward and actually a better developer experience.

Yet again, we'll likely need to improve the developer docs a little bit, with a focus on onboarding newer contributors, which I'm happy to help with in the future.

**Comment by ciampo** [comment]
Slightly late to the conversation !

 @gziolo makes some good points:

- the linter rules would not avoid "mistakes", but rather enforce certain assertions over others
- we'd be leaning onto non-official jest assertions, which some developers may not be used to

But my point of view is very similar to @tyxla 's. I believe that the net result of these refactors (and of enabling the linter rules) is that it is going to be easier for contributors to write tests:

- these assertions make the tests easier to read and understand (while taking also care of some implementation details), following the idea that unit/integration tests can still be written from the point of view of the user
- these new assertions are also pretty easy to learn (especially in a IDE with autocompletion)
- the linter would also guide developers to pick the correct assertion
- finally, I greatly value consistency. If all tests in the repository are written using the same testing framework (jest + RTL + user-event) and use the same style of assertions, it's much easier for a developer to use existing tests as example and quickly understand how a test is supposed to be written

**Comment by gziolo** [comment]
> Slightly late to the conversation !

Actually, it's great timing. My initial idea was to open an issue to have a place to discuss all of the above in addition to the efforts in mentioned PRs to standardize the existing assertions when working with React components. I appreciate all the work in that regard and find it all valuable. My only concern was the mention of the planned integration with linter rules that enforce certain code syntax over anything else. 

I definitely agree with all the points that @tyxla and @ciampo shared. In fact, I probably enforced or helped to enforce many other stricter linting rules in the past. There are two things that I learned over time. While all the tooling is great for a big group of seasoned developers like you two, I also hear the opposite feedback that CI failures triggered by the same tools are sometimes intimidating or hard to address for other contributors. That's why I often try to play Devil's advocate and exercise whether the benefits of introducing stricter validation have enough pros than the potential hassle in the contributors' journey. You make a good job explaining the value of adding those additional rules, which I appreciate very much. I don't have any more objections if it follows with some references in the documentation for testing to make it easier to folks to learn about the recommendations that come with the plugin. 

Aside, at some point, it would be great to share an update at https://make.wordpress.org/core/ with the most recent state of unit testing for components in Gutenberg:
- enzyme removed
- RTL as a recommended way for writing component tests
- a few best practices and maybe references if people want to learn more

It can be timed with React 18 upgrade, as it seems loosely interconnected.

Finally, I would like to share my gratitude to your work @tyxla and @ciampo for working on constant improvements to the developer experience in the Gutenberg project.

**Comment by tyxla** [comment]
Thank you @gziolo, we greatly value a different point of view and we're definitely going to consider p2-ing about this. I'm happy to start a post on make.wp.org/core and get everyone involved to participate. Stay tuned.

File: packages/a11y/src/test/clear.test.js [hunk 1]
```diff
@@ -16,7 +16,7 @@ describe( 'clear', () => {
 		document.body.appendChild( container2 );
 
 		clear();
-		expect( container1.textContent ).toBe( '' );
-		expect( container2.textContent ).toBe( '' );
+		expect( container1 ).toHaveTextContent( '' );
+		expect( container2 ).toHaveTextContent( '' );
```
Comment (jsnajdr) [inline packages/a11y/src/test/clear.test.js:20]: Good morning @tyxla :wave: using the assertion with an empty string is not good. As the assertion checks for substrings by default, it will always match! Fortunately the matcher is smart enough to detect this and when a non-empty string is being matched against `''`:
```js
expect( elWithActualText ).toHaveTextContent( '' );
```
it will fail with some advice:
```
Checking with empty string will always match, use .toBeEmptyDOMElement() instead:
```

Comment (tyxla) [reply packages/a11y/src/test/clear.test.js:20]: That makes sense @jsnajdr, thanks for pointing it out. Will follow-up (if someone hasn't addressed it yet).

Comment (tyxla) [reply packages/a11y/src/test/clear.test.js:20]: Fix here #45007
